cns/fsnotify/fsnotify.go

```diff
@@ -0,0 +1,169 @@
+package fsnotify
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+// map containing containerIDs from failure to release IP
+var releaseIPRetry = make(map[string]string)
+
+type cnsclient interface {
+	ReleaseIPs(ctx context.Context, ipconfig cns.IPConfigsRequest) error
+}
+
+type Watcher struct {
+	CnsClient cnsclient
+	Path      string
+	Logger    *zap.Logger
+}
+
+// WatchFS starts the filesystem watcher to handle async Pod deletes.
+// Blocks until the context is closed; returns underlying fsnotify errors
+// if something goes fatally wrong.
+func (w *Watcher) WatchFs(ctx context.Context) error {
+	// Create new fs watcher.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return errors.Wrap(err, "error creating watcher")
+	}
+	defer watcher.Close()
+
+	c, cancel := context.WithCancel(ctx)
+	// Start listening for events.
+	w.Logger.Info("listening for events from watcher")
+	go func() {
+		defer cancel()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				w.Logger.Info("received event", zap.String("event", event.Name))
+				if event.Has(fsnotify.Create) {
+					w.Logger.Info("file created, triggering release", zap.String("event", event.Name))
+					cnsReleaseErr := w.releaseIP(ctx, event.Name)
+					if cnsReleaseErr != nil {
+						w.Logger.Error("failed to release IP from CNS, adding to map for retry later", zap.Error(cnsReleaseErr))
+						releaseIPRetry[event.Name] = event.Name
+					} else {
+						deleteErr := RemoveFile(event.Name, w.Path)
+						if deleteErr != nil {
+							w.Logger.Error("failed to remove file", zap.Error(err))
+						}
+					}
+				}
+				if event.Has(fsnotify.Remove) {
+					w.Logger.Info("file deleted", zap.String("event", event.Name))
+				}
+			case watcherErr, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				w.Logger.Error("watcher error", zap.Error(watcherErr))
+			}
+		}
+	}()
+
+	// periodically check map for release ip retries
+	// remove entry from map ip is successfully released
+	// on failure the entry will stay in the map
+	ticker := time.NewTicker(15 * time.Second) //nolint
+	stop := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				if len(releaseIPRetry) != 0 {
+					w.Logger.Info("attempt delete on entries in releaseIPRetry")
+					for _, entry := range releaseIPRetry {
+						retryErr := w.releaseIP(ctx, entry)
+						if retryErr != nil {
+							w.Logger.Error("failed to release IP from CNS", zap.Error(retryErr))
+						} else {
+							delete(releaseIPRetry, entry)
+							removeErr := RemoveFile(entry, w.Path)
+							if removeErr != nil {
+								w.Logger.Error("failed to remove file", zap.Error(removeErr))
```

Comment:
what happen if we failed to remove the file from the fs here? I think it would make the operation doesn't feel "atomic" since even though we did successfully ReleaseIP'ed the entry, we can still see the related file in the fs, indicating that it did not get ReleaseIP'ed.